lib/srv/reexec.go

```diff
@@ -653,29 +658,60 @@ func handleRemotePortForward(ctx context.Context, params forwardHandlerParams) e
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	defer controlConn.Close()
 
-	addr, err := net.ResolveTCPAddr("tcp", params.addr)
+	listener, err := net.Listen("tcp", params.addr)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	listener, err := net.ListenTCP("tcp", addr)
-	if err != nil {
+	// Listen may have chosen a new port, so communicate the final address to
+	// the parent process.
+	if _, _, err = controlConn.WriteWithFDs([]byte(listener.Addr().String()), nil); err != nil {
+		listener.Close()
 		return trace.Wrap(err)
 	}
-	defer listener.Close()
-	listenerFD, err := listener.File()
+	params.controlConn = controlConn
+
+	go func() {
+		defer listener.Close()
+
+		for ctx.Err() == nil {
+			conn, err := listener.Accept()
+			if err != nil {
+				log.WithError(err).Error("Remote forwarding listener failed.")
+				return
+			}
+			go func() {
+				if err := forwardAcceptedConn(ctx, conn, params); err != nil {
+					log.WithError(err).Error("Error handling port forwarding connection.")
+				}
+			}()
+		}
+	}()
+
+	return nil
+}
+
+// forwardAcceptedConn forwards the given conn to the parent process over a new
+// socket.
+func forwardAcceptedConn(ctx context.Context, conn net.Conn, params forwardHandlerParams) error {
+	local, remote, err := uds.NewSocketpair(uds.SocketTypeStream)
```

Comment:
If we're ok with sending file descriptors from the child to the parent then there's no reason not to send just the listener - it'd be easier to deal with and even less "problematic" in terms of what a broken child process could mess with.

If we've decided that we are not ok with sending TCP sockets from the child to the parent we shouldn't be ok with sending socketpair halves either; if we have to go for the data copying way then the parent should keep spawning socketpairs and sending them to the child, and the child can then read one for every new accepted connection on the listener - that way no socket is ever used by the parent while the child has a reference to the same socket and only data is passed through.